src/gestures.rs

```diff
@@ -172,142 +203,137 @@ impl Identifiable for PinchGesture {
         }
     }
 }
 
-
 pub struct Listener<'a> {
     swipe: SwipeBuilder,
     pinch: PinchBuilder,
-    gesture_action: &'a dyn Fn(GestureType)
+    gesture_action: &'a dyn Fn(GestureType),
 }
 
 impl<'a> Listener<'a> {
-
     pub fn new(gesture_action: &dyn Fn(GestureType)) -> Listener {
         Listener {
             swipe: SwipeBuilder::empty(),
             pinch: PinchBuilder::empty(),
-            gesture_action
+            gesture_action,
         }
     }
 
     pub fn event(mut self, event: input::Event) -> Self {
-
         match event {
-            Gesture(Swipe(Begin(event))) =>
-                self.swipe.new(event.finger_count()),
-            Gesture(Swipe(Update(event))) =>
-                self.swipe.update(event),
+            Gesture(Swipe(Begin(event))) => self.swipe.new(event.finger_count()),
+            Gesture(Swipe(Update(event))) => {
+                self.swipe.update(event);
+            }
             Gesture(Swipe(End(event))) => {
                 match self.swipe.build(event) {
-                    Ok(g) => {
-                        match g.gesture_type() {
-                            Some(t) => (self.gesture_action)(t),
-                            None => warn!("cancelled or unrecognized gesture {:?}", g)
-                        }
+                    Ok(g) => match g.gesture_type() {
+                        Some(t) => (self.gesture_action)(t),
+                        None => warn!("cancelled or unrecognized gesture {:?}", g),
                     },
-                    Err(s) => error!("no Gesture {:?}", s)
+                    Err(s) => error!("no Gesture {:?}", s),
                 }
                 self.swipe = SwipeBuilder::empty();
+            }
+
+            Gesture(Pinch(GesturePinchEvent::Begin(event))) => self.pinch.new(event.scale()),
+            Gesture(Pinch(GesturePinchEvent::Update(event))) => match self.pinch.update(event) {
+                Ok(p) => match p.gesture_type() {
+                    Some(t) => (self.gesture_action)(t),
+                    None => warn!("cancelled or unrecognized gesture {:?}", p),
+                },
+                Err(s) => error!("no Gesture {:?}", s),
             },
 
-            Gesture(Pinch(GesturePinchEvent::Begin(event))) =>
-                self.pinch.new(event.scale()),
-            Gesture(Pinch(GesturePinchEvent::Update(event))) =>
-                self.pinch.update(event),
             Gesture(Pinch(GesturePinchEvent::End(event))) => {
                 let gesture = self.pinch.build(event);
                 match gesture {
-                    Ok(p) => {
-                        match p.gesture_type() {
-                            Some(t) => (self.gesture_action)(t),
-                            None => warn!("cancelled or unrecognized gesture {:?}", p)
-                        }
+                    Ok(p) => match p.gesture_type() {
+                        Some(t) => (self.gesture_action)(t),
```

Comment:
triggering action on every libinput update works... but is not a smooth approach... also not sure what impact in performance this will have because we are trying to trigger actions many many times per second which i guess is not even very useful in most GUI interactions

so i guess we need some configurability for action triggers - action will be triggered at libinput gesture end or (for the swipe to zoom case) on a certain step